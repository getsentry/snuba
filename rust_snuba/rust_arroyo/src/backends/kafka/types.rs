use rdkafka::message::OwnedHeaders;

#[derive(Clone)]
pub struct KafkaPayload {
    pub key: Option<Vec<u8>>,
    pub headers: Option<OwnedHeaders>,
    pub payload: Option<Vec<u8>>,
}
