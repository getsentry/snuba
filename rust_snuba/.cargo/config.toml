[env]
# Workaround for https://github.com/confluentinc/librdkafka/pull/5012
CMAKE_POLICY_VERSION_MINIMUM = "3.10"
