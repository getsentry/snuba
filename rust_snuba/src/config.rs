use std::collections::HashMap;

use serde::{Deserialize, Deserializer};
use serde_json::Value;

#[derive(Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct ConsumerConfig {
    pub storages: Vec<StorageConfig>,
    pub raw_topic: TopicConfig,
    pub commit_log_topic: Option<TopicConfig>,
    pub replacements_topic: Option<TopicConfig>,
    pub dlq_topic: Option<TopicConfig>,
    pub max_batch_size: usize,
    pub max_batch_time_ms: u64,
    pub env: EnvConfig,
}

pub fn deserialize_broker_config<'de, D>(
    deserializer: D,
) -> Result<HashMap<String, String>, D::Error>
where
    D: Deserializer<'de>,
{
    let data = RawBrokerConfig::deserialize(deserializer)?
        .iter()
        .filter_map(|(k, v)| {
            if v.is_null() {
                None
            } else if v.is_number() {
                // Numeric types are valid in confluent-kafka-python config but not in the Rust library
                Some((k.to_string(), v.as_number().unwrap().to_string()))
            } else if v.is_string() {
                if v.as_str().unwrap().is_empty() {
                    return None;
                }
                Some((k.to_string(), v.as_str().unwrap().to_string()))
            } else {
                panic!("Unsupported type");
            }
        })
        .collect();

    Ok(data)
}

#[derive(Deserialize, Debug)]
pub struct TopicConfig {
    pub physical_topic_name: String,
    pub logical_topic_name: String,
    #[serde(deserialize_with = "deserialize_broker_config")]
    pub broker_config: BrokerConfig,
}

type RawBrokerConfig = HashMap<String, Value>;

pub type BrokerConfig = HashMap<String, String>;

impl ConsumerConfig {
    pub fn load_from_str(payload: &str) -> Result<Self, anyhow::Error> {
        let d: Self = serde_json::from_str(payload)?;
        Ok(d)
    }
}

#[derive(Deserialize, Clone, Debug)]
#[serde(deny_unknown_fields)]
pub struct StorageConfig {
    pub name: String,
    pub clickhouse_table_name: String,
    pub clickhouse_cluster: ClickhouseConfig,
    pub message_processor: MessageProcessorConfig,
}

#[derive(Deserialize, Clone, Debug)]
#[serde(deny_unknown_fields)]
pub struct ClickhouseConfig {
    pub host: String,
    pub port: u16,
    pub http_port: u16,
    pub user: String,
    pub password: String,
    pub database: String,
}

#[derive(Deserialize, Clone, Debug)]
#[serde(deny_unknown_fields)]
pub struct MessageProcessorConfig {
    pub python_class_name: String,
    pub python_module: String,
}

#[derive(Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct EnvConfig {
    pub sentry_dsn: Option<String>,
    pub dogstatsd_host: Option<String>,
    pub dogstatsd_port: Option<u16>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_config() {
        let raw =
            "{\"physical_topic_name\": \"test\", \"logical_topic_name\": \"test\", \"broker_config\": {\"bootstrap.servers\": \"127.0.0.1:9092\", \"queued.max.messages.kbytes\": 10000}}";

        let topic_config: TopicConfig = serde_json::from_str(raw).unwrap();

        assert_eq!(
            topic_config.broker_config["queued.max.messages.kbytes"],
            "10000"
        );
    }
}
