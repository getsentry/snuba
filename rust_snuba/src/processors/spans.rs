use crate::types::{BytesInsertBatch, KafkaMessageMetadata};
use rust_arroyo::backends::kafka::types::KafkaPayload;
use rust_arroyo::processing::strategies::InvalidMessage;
use serde::{Deserialize, Deserializer, Serialize};
use std::collections::BTreeMap;
use uuid::Uuid;

pub fn process_message(
    payload: KafkaPayload,
    _metadata: KafkaMessageMetadata,
) -> Result<BytesInsertBatch, InvalidMessage> {
    if let Some(payload_bytes) = payload.payload {
        let msg: FromSpanMessage = serde_json::from_slice(&payload_bytes).map_err(|err| {
            log::error!("Failed to deserialize message: {}", err);
            InvalidMessage
        })?;
        let mut span: Span = msg.try_into()?;

        span.offset = _metadata.offset;
        span.partition = _metadata.partition;

        let serialized = serde_json::to_vec(&span).map_err(|err| {
            log::error!("Failed to serialize processed message: {}", err);
            InvalidMessage
        })?;

        return Ok(BytesInsertBatch {
            rows: vec![serialized],
        });
    }
    Err(InvalidMessage)
}

#[derive(Debug, Default, Deserialize)]
struct FromSpanMessage {
    #[serde(default)]
    description: String,
    duration_ms: u32,
    event_id: Uuid,
    exclusive_time_ms: f64,
    #[serde(deserialize_with = "hex_to_u64")]
    group_raw: u64,
    is_segment: bool,
    #[serde(deserialize_with = "hex_to_u64")]
    parent_span_id: u64,
    project_id: u64,
    retention_days: u16,
    #[serde(deserialize_with = "hex_to_u64")]
    segment_id: u64,
    sentry_tags: FromSentryTags,
    #[serde(deserialize_with = "hex_to_u64")]
    span_id: u64,
    start_timestamp_ms: u64,
    tags: BTreeMap<String, String>,
    trace_id: Uuid,
}

#[derive(Debug, Default, Deserialize)]
struct FromSentryTags {
    #[serde(default)]
    action: String,
    #[serde(default)]
    domain: String,
    #[serde(default)]
    group: String,
    #[serde(rename(deserialize = "http.method"))]
    http_method: Option<String>,
    #[serde(default)]
    module: String,
    #[serde(default)]
    op: String,
    #[serde(default)]
    status: SpanStatus,
    status_code: Option<String>,
    #[serde(default)]
    system: String,
    #[serde(default)]
    transaction: String,
    #[serde(rename(deserialize = "transaction.method"))]
    transaction_method: Option<String>,
    #[serde(default, rename(deserialize = "transaction.op"))]
    transaction_op: String,
}

impl FromSentryTags {
    fn to_keys_values(&self) -> (Vec<String>, Vec<String>) {
        let mut tags: BTreeMap<String, String> = BTreeMap::new();

        tags.insert("action".into(), self.action.clone());
        tags.insert("domain".into(), self.domain.clone());
        tags.insert("group".into(), self.group.clone());
        tags.insert("module".into(), self.module.clone());
        tags.insert("op".into(), self.op.clone());
        tags.insert("status".into(), self.status.as_str().to_string());
        tags.insert("system".into(), self.system.clone());
        tags.insert("transaction".into(), self.transaction.clone());
        tags.insert("transaction.op".into(), self.transaction_op.clone());

        if let Some(http_method) = &self.http_method {
            tags.insert("http.method".into(), http_method.into());
        }

        if let Some(transaction_method) = &self.transaction_method {
            tags.insert("transaction.method".into(), transaction_method.into());
        }

        if let Some(status_code) = &self.status_code {
            tags.insert("status_code".into(), status_code.into());
        }

        (
            tags.keys().cloned().collect(),
            tags.values().cloned().collect(),
        )
    }
}

#[derive(Debug, Default, Serialize)]
struct Span {
    #[serde(default)]
    action: String,
    deleted: u8,
    description: String,
    #[serde(default)]
    domain: String,
    duration: u32,
    end_ms: u16,
    end_timestamp: u64,
    exclusive_time: f64,
    group: u64,
    group_raw: u64,
    is_segment: u8,
    #[serde(default)]
    module: String,
    #[serde(rename(serialize = "measurements.key"))]
    measurement_keys: Vec<String>,
    #[serde(rename(serialize = "measurements.value"))]
    measurement_values: Vec<f64>,
    offset: u64,
    op: String,
    parent_span_id: u64,
    partition: u16,
    platform: String,
    project_id: u64,
    retention_days: u16,
    segment_id: u64,
    segment_name: String,
    #[serde(rename(serialize = "sentry_tags.key"))]
    sentry_tag_keys: Vec<String>,
    #[serde(rename(serialize = "sentry_tags.value"))]
    sentry_tag_values: Vec<String>,
    span_id: u64,
    span_kind: String,
    span_status: u8,
    start_ms: u16,
    start_timestamp: u64,
    status: u32,
    #[serde(rename(serialize = "tags.key"))]
    tag_keys: Vec<String>,
    #[serde(rename(serialize = "tags.value"))]
    tag_values: Vec<String>,
    trace_id: Uuid,
    transaction_id: Uuid,
    transaction_op: String,
    user: String,
}

impl TryFrom<FromSpanMessage> for Span {
    type Error = InvalidMessage;

    fn try_from(from: FromSpanMessage) -> Result<Span, InvalidMessage> {
        let end_timestamp_ms = from.start_timestamp_ms + from.duration_ms as u64;
        let status = from.sentry_tags.status as u8;
        let transaction_op = from.sentry_tags.transaction_op.clone();
        let (sentry_tag_keys, sentry_tag_values) = from.sentry_tags.to_keys_values();
        let mut tag_keys: Vec<String> = from.tags.clone().into_keys().collect();
        let mut tag_values: Vec<String> = from.tags.into_values().collect();

        if let Some(http_method) = from.sentry_tags.http_method.clone() {
            tag_keys.push("http.method".into());
            tag_values.push(http_method);
        }

        if let Some(status_code) = &from.sentry_tags.status_code {
            tag_keys.push("status_code".into());
            tag_values.push(status_code.into());
        }

        if let Some(transaction_method) = from.sentry_tags.transaction_method.clone() {
            tag_keys.push("transaction.method".into());
            tag_values.push(transaction_method);
        }

        Ok(Self {
            action: from.sentry_tags.action.clone(),
            description: from.description,
            domain: from.sentry_tags.domain.clone(),
            duration: from.duration_ms,
            end_ms: (end_timestamp_ms % 1000) as u16,
            end_timestamp: end_timestamp_ms / 1000,
            exclusive_time: from.exclusive_time_ms,
            group: u64::from_str_radix(&from.sentry_tags.group, 16).map_err(|_| InvalidMessage)?,
            group_raw: from.group_raw,
            is_segment: if from.is_segment { 1 } else { 0 },
            module: from.sentry_tags.module.clone(),
            op: from.sentry_tags.op.clone(),
            parent_span_id: from.parent_span_id,
            platform: from.sentry_tags.system.clone(),
            project_id: from.project_id,
            retention_days: from.retention_days,
            segment_id: from.segment_id,
            segment_name: from.sentry_tags.transaction.clone(),
            sentry_tag_keys,
            sentry_tag_values,
            span_id: from.span_id,
            span_status: status,
            start_ms: (from.start_timestamp_ms % 1000) as u16,
            start_timestamp: from.start_timestamp_ms / 1000,
            status: status.into(),
            tag_keys,
            tag_values,
            trace_id: from.trace_id,
            transaction_id: from.event_id,
            transaction_op,
            ..Default::default()
        })
    }
}

fn hex_to_u64<'de, D>(deserializer: D) -> Result<u64, D::Error>
where
    D: Deserializer<'de>,
{
    let hex = String::deserialize(deserializer)?;
    u64::from_str_radix(&hex, 16).map_err(serde::de::Error::custom)
}

#[derive(Clone, Copy, Default, Debug, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
#[repr(u8)] // size limit in clickhouse
pub enum SpanStatus {
    /// The operation completed successfully.
    ///
    /// HTTP status 100..299 + successful redirects from the 3xx range.
    Ok = 0,

    /// The operation was cancelled (typically by the user).
    Cancelled = 1,

    /// Unknown. Any non-standard HTTP status code.
    ///
    /// "We do not know whether the transaction failed or succeeded"
    #[default]
    #[serde(alias = "")]
    Unknown = 2,

    /// Client specified an invalid argument. 4xx.
    ///
    /// Note that this differs from FailedPrecondition. InvalidArgument indicates arguments that
    /// are problematic regardless of the state of the system.
    InvalidArgument = 3,

    /// Deadline expired before operation could complete.
    ///
    /// For operations that change the state of the system, this error may be returned even if the
    /// operation has been completed successfully.
    ///
    /// HTTP redirect loops and 504 Gateway Timeout
    DeadlineExceeded = 4,

    /// 404 Not Found. Some requested entity (file or directory) was not found.
    NotFound = 5,

    /// Already exists (409)
    ///
    /// Some entity that we attempted to create already exists.
    AlreadyExists = 6,

    /// 403 Forbidden
    ///
    /// The caller does not have permission to execute the specified operation.
    PermissionDenied = 7,

    /// 429 Too Many Requests
    ///
    /// Some resource has been exhausted, perhaps a per-user quota or perhaps the entire file
    /// system is out of space.
    ResourceExhausted = 8,

    /// Operation was rejected because the system is not in a state required for the operation's
    /// execution
    FailedPrecondition = 9,

    /// The operation was aborted, typically due to a concurrency issue.
    Aborted = 10,

    /// Operation was attempted past the valid range.
    OutOfRange = 11,

    /// 501 Not Implemented
    ///
    /// Operation is not implemented or not enabled.
    Unimplemented = 12,

    /// Other/generic 5xx.
    InternalError = 13,

    /// 503 Service Unavailable
    Unavailable = 14,

    /// Unrecoverable data loss or corruption
    DataLoss = 15,

    /// 401 Unauthorized (actually does mean unauthenticated according to RFC 7235)
    ///
    /// Prefer PermissionDenied if a user is logged in.
    Unauthenticated = 16,
}

impl SpanStatus {
    /// Returns the string representation of the status.
    pub fn as_str(&self) -> &'static str {
        match *self {
            SpanStatus::Ok => "ok",
            SpanStatus::DeadlineExceeded => "deadline_exceeded",
            SpanStatus::Unauthenticated => "unauthenticated",
            SpanStatus::PermissionDenied => "permission_denied",
            SpanStatus::NotFound => "not_found",
            SpanStatus::ResourceExhausted => "resource_exhausted",
            SpanStatus::InvalidArgument => "invalid_argument",
            SpanStatus::Unimplemented => "unimplemented",
            SpanStatus::Unavailable => "unavailable",
            SpanStatus::InternalError => "internal_error",
            SpanStatus::Unknown => "unknown",
            SpanStatus::Cancelled => "cancelled",
            SpanStatus::AlreadyExists => "already_exists",
            SpanStatus::FailedPrecondition => "failed_precondition",
            SpanStatus::Aborted => "aborted",
            SpanStatus::OutOfRange => "out_of_range",
            SpanStatus::DataLoss => "data_loss",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::DateTime;
    use rust_arroyo::backends::kafka::types::KafkaPayload;
    use std::time::SystemTime;

    #[test]
    fn test_valid_span() {
        let data = r#"{
            "duration_ms": 1000,
            "event_id": "dcc403b73ef548648188bbfa6012e9dc",
            "exclusive_time_ms": 1000,
            "group_raw": "b640a0ce465fa2a4",
            "is_segment": false,
            "organization_id": 69,
            "parent_span_id": "deadbeefdeadbeef",
            "project_id": 1,
            "retention_days": 90,
            "segment_id": "deadbeefdeadbeef",
            "span_id": "deadbeefdeadbeef",
            "start_timestamp_ms": 1691105878720,
            "trace_id": "deadbeefdeadbeefdeadbeefdeadbeef",
            "tags": {
              "tag1": "value1",
              "tag2": "123",
              "tag3": "true"
            },
            "sentry_tags": {
              "action": "GET",
              "domain": "targetdomain.tld:targetport",
              "group": "deadbeefdeadbeef",
              "http.method": "GET",
              "module": "http",
              "op": "http.client",
              "status": "ok",
              "status_code": "200",
              "system": "python",
              "transaction": "/organizations/:orgId/issues/",
              "transaction.method": "GET",
              "transaction.op": "navigation"
            }
          }"#;
        let payload = KafkaPayload {
            key: None,
            headers: None,
            payload: Some(data.as_bytes().to_vec()),
        };
        let meta = KafkaMessageMetadata {
            partition: 0,
            offset: 1,
            timestamp: DateTime::from(SystemTime::now()),
        };
        process_message(payload, meta).expect("The message should be processed");
    }

    #[test]
    fn test_no_status_value() {
        let data = r#"{
            "duration_ms": 1000,
            "event_id": "dcc403b73ef548648188bbfa6012e9dc",
            "exclusive_time_ms": 1000,
            "group_raw": "b640a0ce465fa2a4",
            "is_segment": false,
            "organization_id": 69,
            "parent_span_id": "deadbeefdeadbeef",
            "project_id": 1,
            "retention_days": 90,
            "segment_id": "deadbeefdeadbeef",
            "span_id": "deadbeefdeadbeef",
            "start_timestamp_ms": 1691105878720,
            "trace_id": "deadbeefdeadbeefdeadbeefdeadbeef",
            "tags": {
              "tag1": "value1",
              "tag2": "123",
              "tag3": "true"
            },
            "sentry_tags": {
              "action": "GET",
              "domain": "targetdomain.tld:targetport",
              "group": "deadbeefdeadbeef",
              "http.method": "GET",
              "module": "http",
              "op": "http.client",
              "status_code": "200",
              "system": "python",
              "transaction": "/organizations/:orgId/issues/",
              "transaction.method": "GET",
              "transaction.op": "navigation"
            }
          }"#;
        let payload = KafkaPayload {
            key: None,
            headers: None,
            payload: Some(data.as_bytes().to_vec()),
        };
        let meta = KafkaMessageMetadata {
            partition: 0,
            offset: 1,
            timestamp: DateTime::from(SystemTime::now()),
        };
        process_message(payload, meta).expect("The message should be processed");
    }

    #[test]
    fn test_empty_status_value() {
        let data = r#"{
            "duration_ms": 1000,
            "event_id": "dcc403b73ef548648188bbfa6012e9dc",
            "exclusive_time_ms": 1000,
            "group_raw": "b640a0ce465fa2a4",
            "is_segment": false,
            "organization_id": 69,
            "parent_span_id": "deadbeefdeadbeef",
            "project_id": 1,
            "retention_days": 90,
            "segment_id": "deadbeefdeadbeef",
            "span_id": "deadbeefdeadbeef",
            "start_timestamp_ms": 1691105878720,
            "trace_id": "deadbeefdeadbeefdeadbeefdeadbeef",
            "tags": {
              "tag1": "value1",
              "tag2": "123",
              "tag3": "true"
            },
            "sentry_tags": {
              "action": "GET",
              "domain": "targetdomain.tld:targetport",
              "group": "deadbeefdeadbeef",
              "http.method": "GET",
              "module": "http",
              "op": "http.client",
              "status": "",
              "status_code": "200",
              "system": "python",
              "transaction": "/organizations/:orgId/issues/",
              "transaction.method": "GET",
              "transaction.op": "navigation"
            }
          }"#;
        let payload = KafkaPayload {
            key: None,
            headers: None,
            payload: Some(data.as_bytes().to_vec()),
        };
        let meta = KafkaMessageMetadata {
            partition: 0,
            offset: 1,
            timestamp: DateTime::from(SystemTime::now()),
        };
        process_message(payload, meta).expect("The message should be processed");
    }
}
