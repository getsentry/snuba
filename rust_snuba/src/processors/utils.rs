use serde::{Deserializer, Deserialize};

pub const DEFAULT_RETENTION_DAYS: u16 = 90;

pub fn default_retention_days() -> Option<u16> {
    Some(DEFAULT_RETENTION_DAYS)
}

pub fn hex_to_u64<'de, D>(deserializer: D) -> Result<u64, D::Error>
where
    D: Deserializer<'de>,
{
    let hex = String::deserialize(deserializer)?;
    u64::from_str_radix(&hex, 16).map_err(serde::de::Error::custom)
}
