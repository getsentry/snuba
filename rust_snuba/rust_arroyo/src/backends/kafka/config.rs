use rdkafka::config::ClientConfig as RdKafkaConfig;
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct KafkaConfig {
    config_map: HashMap<String, String>,
}

impl KafkaConfig {
    pub fn new_config(
        bootstrap_servers: Vec<String>,
        override_params: Option<HashMap<String, String>>,
    ) -> Self {
        let mut config_map = HashMap::new();
        config_map.insert("bootstrap.servers".to_string(), bootstrap_servers.join(","));

        let config = Self { config_map };

        apply_override_params(config, override_params)
    }

    pub fn new_consumer_config(
        bootstrap_servers: Vec<String>,
        group_id: String,
        auto_offset_reset: String,
        _strict_offset_reset: bool, // TODO: Implement this
        override_params: Option<HashMap<String, String>>,
    ) -> Self {
        let mut config = KafkaConfig::new_config(bootstrap_servers, None);
        config.config_map.insert("group.id".to_string(), group_id);
        config
            .config_map
            .insert("enable.auto.commit".to_string(), "false".to_string());
        config
            .config_map
            .insert("auto.offset.reset".to_string(), auto_offset_reset);

        apply_override_params(config, override_params)
    }

    pub fn new_producer_config(
        bootstrap_servers: Vec<String>,
        override_params: Option<HashMap<String, String>>,
    ) -> Self {
        let config = KafkaConfig::new_config(bootstrap_servers, None);

        apply_override_params(config, override_params)
    }
}

impl From<KafkaConfig> for RdKafkaConfig {
    fn from(item: KafkaConfig) -> Self {
        let mut config_obj = RdKafkaConfig::new();
        for (key, val) in item.config_map.iter() {
            config_obj.set(key, val);
        }
        config_obj
    }
}

fn apply_override_params(
    mut config: KafkaConfig,
    override_params: Option<HashMap<String, String>>,
) -> KafkaConfig {
    match override_params {
        Some(params) => {
            for (param, value) in params {
                config.config_map.insert(param, value);
            }
        }
        None => {}
    }
    config
}

#[cfg(test)]
mod tests {
    use super::KafkaConfig;
    use rdkafka::config::ClientConfig as RdKafkaConfig;
    use std::collections::HashMap;

    #[test]
    fn test_build_consumer_configuration() {
        let config = KafkaConfig::new_consumer_config(
            vec!["localhost:9092".to_string()],
            "my-group".to_string(),
            "error".to_string(),
            false,
            Some(HashMap::from([(
                "queued.max.messages.kbytes".to_string(),
                "1000000".to_string(),
            )])),
        );

        let rdkafka_config: RdKafkaConfig = config.into();
        assert_eq!(
            rdkafka_config.get("queued.max.messages.kbytes"),
            Some("1000000")
        );
    }
}
