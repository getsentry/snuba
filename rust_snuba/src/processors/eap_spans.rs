use anyhow::Context;
use chrono::DateTime;
use seq_macro::seq;
use serde::Serialize;
use std::collections::HashMap;
use uuid::Uuid;

use rust_arroyo::backends::kafka::types::KafkaPayload;

use crate::config::ProcessorConfig;
use crate::processors::spans::FromSpanMessage;
use crate::processors::utils::enforce_retention;
use crate::types::{InsertBatch, KafkaMessageMetadata};

pub fn process_message(
    payload: KafkaPayload,
    _metadata: KafkaMessageMetadata,
    config: &ProcessorConfig,
) -> anyhow::Result<InsertBatch> {
    let payload_bytes = payload.payload().context("Expected payload")?;
    let msg: FromSpanMessage = serde_json::from_slice(payload_bytes)?;

    let origin_timestamp = DateTime::from_timestamp(msg.received as i64, 0);
    let mut span: EAPSpan = msg.try_into()?;

    span.retention_days = Some(enforce_retention(span.retention_days, &config.env_config));

    InsertBatch::from_rows([span], origin_timestamp)
}

seq!(N in 0..200 {
#[derive(Debug, Default, Serialize)]
struct EAPSpan {
    // the span object for the new "events analytics platform"
    organization_id: u64,
    project_id: u64,
    service: String, //currently just project ID as a string
    trace_id: Uuid,
    span_id: u64,
    #[serde(default)]
    parent_span_id: u64,
    segment_id: u64,      //aka transaction ID
    segment_name: String, //aka transaction name
    is_segment: bool,     //aka "is transaction"
    start_timestamp: u64,
    end_timestamp: u64,
    duration_ms: u32,
    exclusive_time_ms: f64,
    retention_days: Option<u16>,
    name: String, //aka description

    sampling_factor: f64,
    sampling_weight: f64,
    sign: u8, //1 for additions, -1 for deletions - for this worker it should be 1

    #(
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    attr_str_~N: HashMap<String, String>,

    #[serde(skip_serializing_if = "HashMap::is_empty")]
    attr_num_~N: HashMap<String, f64>,
    )*
}
});

fn fnv_1a(input: &[u8]) -> u32 {
    const FNV_1A_PRIME: u32 = 16777619;
    const FNV_1A_OFFSET_BASIS: u32 = 2166136261;

    let mut res = FNV_1A_OFFSET_BASIS;
    for byt in input {
        res ^= *byt as u32;
        res = res.wrapping_mul(FNV_1A_PRIME);
    }

    res
}

impl From<FromSpanMessage> for EAPSpan {
    fn from(from: FromSpanMessage) -> EAPSpan {
        let sentry_tags = from.sentry_tags.unwrap_or_default();
        let tags = from.tags.unwrap_or_default();
        let measurements = from.measurements.unwrap_or_default();

        let mut res = Self {
            organization_id: from.organization_id,
            project_id: from.project_id,
            trace_id: from.trace_id,
            service: format!("{:x}", from.project_id),
            span_id: u64::from_str_radix(&from.span_id, 16).unwrap_or_default(),
            parent_span_id: from
                .parent_span_id
                .map_or(0, |s| u64::from_str_radix(&s, 16).unwrap_or(0)),
            segment_id: from
                .segment_id
                .map_or(0, |s| u64::from_str_radix(&s, 16).unwrap_or(0)),
            segment_name: sentry_tags.get("transaction").cloned().unwrap_or_default(),
            is_segment: from.is_segment,
            start_timestamp: (from.start_timestamp_precise * 1e6) as u64,
            end_timestamp: (from.end_timestamp_precise * 1e6) as u64,
            duration_ms: from.duration_ms,
            exclusive_time_ms: from.exclusive_time_ms,
            retention_days: from.retention_days,
            name: from.description.unwrap_or_default(),

            sampling_weight: 1.,
            sampling_factor: 1.,
            sign: 1,

            ..Default::default()
        };

        {
            seq!(N in 0..200 {
            let mut attr_str_buckets = [
                #(
                &mut res.attr_str_~N,
                )*
            ];
            });

            sentry_tags.iter().chain(tags.iter()).for_each(|(k, v)| {
                attr_str_buckets[(fnv_1a(k.as_bytes()) as usize) % attr_str_buckets.len()]
                    .insert(k.clone(), v.clone());
            });
        }

        {
            seq!(N in 0..200 {
            let mut attr_num_buckets = [
                #(
                &mut res.attr_num_~N,
                )*
            ];
                });

            measurements.iter().for_each(|(k, v)| {
                attr_num_buckets[(fnv_1a(k.as_bytes()) as usize) % attr_num_buckets.len()]
                    .insert(k.clone(), v.value);
            });
        }

        res
    }
}

#[cfg(test)]
mod tests {
    use std::time::SystemTime;

    use super::*;

    const SPAN_KAFKA_MESSAGE: &str = r#"
{
    "description": "/api/0/relays/projectconfigs/",
    "duration_ms": 152,
    "event_id": "d826225de75d42d6b2f01b957d51f18f",
    "exclusive_time_ms": 0.228,
    "is_segment": true,
    "data": {
        "sentry.environment": "development",
        "sentry.release": "backend@24.7.0.dev0+c45b49caed1e5fcbf70097ab3f434b487c359b6b",
        "thread.name": "uWSGIWorker1Core0",
        "thread.id": "8522009600",
        "sentry.segment.name": "/api/0/relays/projectconfigs/",
        "sentry.sdk.name": "sentry.python.django",
        "sentry.sdk.version": "2.7.0"
    },
    "measurements": {
        "num_of_spans": {
            "value": 50.0
        }
    },
    "organization_id": 1,
    "origin": "auto.http.django",
    "project_id": 1,
    "received": 1721319572.877828,
    "retention_days": 90,
    "segment_id": "8873a98879faf06d",
    "sentry_tags": {
        "category": "http",
        "environment": "development",
        "op": "http.server",
        "platform": "python",
        "release": "backend@24.7.0.dev0+c45b49caed1e5fcbf70097ab3f434b487c359b6b",
        "sdk.name": "sentry.python.django",
        "sdk.version": "2.7.0",
        "status": "ok",
        "status_code": "200",
        "thread.id": "8522009600",
        "thread.name": "uWSGIWorker1Core0",
        "trace.status": "ok",
        "transaction": "/api/0/relays/projectconfigs/",
        "transaction.method": "POST",
        "transaction.op": "http.server",
        "user": "ip:127.0.0.1"
    },
    "span_id": "8873a98879faf06d",
    "tags": {
        "http.status_code": "200",
        "relay_endpoint_version": "3",
        "relay_id": "88888888-4444-4444-8444-cccccccccccc",
        "relay_no_cache": "False",
        "relay_protocol_version": "3",
        "relay_use_post_or_schedule": "True",
        "relay_use_post_or_schedule_rejected": "version",
        "server_name": "D23CXQ4GK2.local",
        "spans_over_limit": "False"
    },
    "trace_id": "d099bf9ad5a143cf8f83a98081d0ed3b",
    "start_timestamp_ms": 1721319572616,
    "start_timestamp_precise": 1721319572.616648,
    "end_timestamp_precise": 1721319572.768806
}
    "#;

    #[test]
    fn test_half_md5() {
        //select halfMD5('test') == 688887797400064883
        assert_eq!(fnv_1a("test".as_bytes()), 2949673445)
    }

    #[test]
    fn test_valid_span() {
        let payload = KafkaPayload::new(None, None, Some(SPAN_KAFKA_MESSAGE.as_bytes().to_vec()));
        let meta = KafkaMessageMetadata {
            partition: 0,
            offset: 1,
            timestamp: DateTime::from(SystemTime::now()),
        };
        process_message(payload, meta, &ProcessorConfig::default())
            .expect("The message should be processed");
    }

    #[test]
    fn test_serialization() {
        let msg: FromSpanMessage = serde_json::from_slice(SPAN_KAFKA_MESSAGE.as_bytes()).unwrap();
        let span: EAPSpan = msg.try_into().unwrap();
        insta::with_settings!({sort_maps => true}, {
            insta::assert_json_snapshot!(span)
        });
    }
}
