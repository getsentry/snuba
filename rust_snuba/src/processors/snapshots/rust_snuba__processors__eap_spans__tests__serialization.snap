---
source: src/processors/eap_spans.rs
expression: span
---
{
  "_sort_timestamp": 1721319572,
  "attr_num_1": {
    "my.neg.float.field": -101.2,
    "my.true.bool.field": 1.0
  },
  "attr_num_11": {
    "num_of_spans": 50.0
  },
  "attr_num_14": {
    "my.int.field": 2000.0
  },
  "attr_num_17": {
    "my.float.field": 101.2
  },
  "attr_num_5": {
    "my.neg.field": -100.0
  },
  "attr_num_6": {
    "my.false.bool.field": 0.0
  },
  "attr_str_0": {
    "relay_protocol_version": "3"
  },
  "attr_str_1": {
    "sentry.thread.name": "uWSGIWorker1Core0"
  },
  "attr_str_10": {
    "sentry.sdk.version": "2.7.0"
  },
  "attr_str_11": {
    "sentry.platform": "python",
    "sentry.transaction.method": "POST",
    "sentry.user": "ip:127.0.0.1"
  },
  "attr_str_12": {
    "relay_use_post_or_schedule_rejected": "version",
    "server_name": "D23CXQ4GK2.local"
  },
  "attr_str_14": {
    "sentry.status": "ok"
  },
  "attr_str_17": {
    "relay_endpoint_version": "3",
    "relay_no_cache": "False",
    "relay_use_post_or_schedule": "True",
    "spans_over_limit": "False"
  },
  "attr_str_18": {
    "sentry.segment.name": "/api/0/relays/projectconfigs/",
    "sentry.status_code": "200"
  },
  "attr_str_19": {
    "sentry.op": "http.server"
  },
  "attr_str_3": {
    "sentry.thread.id": "8522009600"
  },
  "attr_str_4": {
    "thread.id": "8522009600"
  },
  "attr_str_5": {
    "http.status_code": "200",
    "sentry.release": "backend@24.7.0.dev0+c45b49caed1e5fcbf70097ab3f434b487c359b6b",
    "sentry.sdk.name": "sentry.python.django",
    "sentry.transaction.op": "http.server"
  },
  "attr_str_6": {
    "relay_id": "88888888-4444-4444-8444-cccccccccccc",
    "thread.name": "uWSGIWorker1Core0"
  },
  "attr_str_7": {
    "sentry.trace.status": "ok"
  },
  "attr_str_8": {
    "sentry.category": "http"
  },
  "attr_str_9": {
    "sentry.environment": "development"
  },
  "duration_micro": 152158,
  "end_timestamp": 1721319572768806,
  "exclusive_time_micro": 228,
  "is_segment": true,
  "name": "/api/0/relays/projectconfigs/",
  "organization_id": 1,
  "parent_span_id": 0,
  "project_id": 1,
  "retention_days": 90,
  "sampling_factor": 0.02,
  "sampling_weight": 50,
  "segment_id": 9832388815107059821,
  "segment_name": "/api/0/relays/projectconfigs/",
  "service": "1",
  "sign": 1,
  "span_id": 9832388815107059821,
  "start_timestamp": 1721319572616648,
  "trace_id": "d099bf9a-d5a1-43cf-8f83-a98081d0ed3b"
}
